phylo/src/substitution_models/mod.rs

```diff
@@ -104,39 +80,140 @@ where
+        let mut probs = SubstitutionModel::get_stationary_distribution(self)
+            .clone()
+            .component_mul(char_encoding);
+        probs.scale_mut(1.0 / probs.sum());
+        probs
+    }
+
+    fn index(&self) -> &[usize; 255] {
+        SubstitutionModel::index(self)
+    }
+
+    fn get_params(&self) -> &Self::Params {
+        &self.params
     }
 }
 
-impl<'a> SubstitutionLikelihoodCost<'a, 20> {
-    pub fn compute_log_likelihood(
+impl<Params: EvoModelParams> ParsimonyModel for SubstModel<Params>
+where
+    SubstModel<Params>: SubstitutionModel,
+{
+    fn generate_scorings(
         &self,
-        model: &ProteinSubstModel,
-    ) -> (f64, SubstitutionModelInfo<20>) {
-        let mut tmp_info = SubstitutionModelInfo::<20>::new(self.info, model).unwrap();
-        let logl = self.compute_log_likelihood_with_tmp(model, &mut tmp_info);
+        times: &[f64],
+        zero_diag: bool,
+        rounding: &Rounding,
+    ) -> HashMap<OrderedFloat<f64>, (SubstMatrix, f64)> {
+        SubstitutionModel::generate_scorings(self, times, zero_diag, rounding)
+    }
+
+    fn get_scoring_matrix(&self, time: f64, rounding: &Rounding) -> (SubstMatrix, f64) {
+        SubstitutionModel::get_scoring_matrix(self, time, rounding)
+    }
+}
+
+#[derive(Clone)]
+pub struct SubstitutionLikelihoodCost<'a, SubstModel: SubstitutionModel + 'a> {
+    pub(crate) info: &'a PhyloInfo,
+    pub(crate) model: &'a SubstModel,
+}
+
+impl<'a, SubstModel: SubstitutionModel + 'a> SubstitutionLikelihoodCost<'a, SubstModel> {
+    pub fn new(info: &'a PhyloInfo, model: &'a SubstModel) -> Self {
+        SubstitutionLikelihoodCost { info, model }
+    }
+    pub fn compute_log_likelihood(&self) -> (f64, SubstModelInfo<SubstModel>) {
+        let mut tmp_info = SubstModelInfo::<SubstModel>::new(self.info, self.model).unwrap();
+        let logl = self.compute_log_likelihood_with_tmp(self.model, &mut tmp_info);
         (logl, tmp_info)
     }
 }
 
-impl<'a, const N: usize> SubstitutionLikelihoodCost<'a, N>
-where
-    Const<N>: DimMin<Const<N>, Output = Const<N>>,
+impl<'a, SubstModel: SubstitutionModel + 'a> LikelihoodCostFunction<'a>
+    for SubstitutionLikelihoodCost<'a, SubstModel>
 {
+    type Model = SubstModel;
+    type Info = SubstModelInfo<SubstModel>;
+
+    fn compute_log_likelihood(&self) -> f64 {
+        self.compute_log_likelihood().0
+    }
+
+    fn get_empirical_frequencies(&self) -> FreqVector {
+        let all_counts = self.info.get_counts();
+        let mut total = all_counts.values().sum::<f64>();
+        let index = SubstitutionModel::index(self.model);
+        let mut freqs = FreqVector::from_column_slice(&vec![0.0; Self::Model::N]);
+        for (&char, &count) in all_counts.iter() {
+            freqs += &Self::Model::char_sets()[char as usize].scale(count);
+        }
+        for &char in Self::Model::ALPHABET {
+            if freqs[index[char as usize]] == 0.0 {
```

Comment:
eq comparison to 0.0 is dangerous. just reminding! you probably want to put a comment in case you really compare to zero values. it might even make sense to compare to f64::Default() just to be explicit